onnx/defs/nn/defs.cc

```diff
@@ -2487,4 +2487,206 @@ ONNX_OPERATOR_SET_SCHEMA(
         )ONNX"
         ));
 
+void col2imShapeInference(InferenceContext& ctx) {
+  propagateElemTypeFromInputToOutput(ctx, 0, 0);
+  // we need at least three inputs to have a shape for this inference.
+  if (!hasNInputShapes(ctx, 3)) {
+    return;
+  }
+
+  // validating 'input' input
+  auto input_shape = ctx.getInputType(0)->tensor_type().shape();
+  if ((input_shape.dim_size() < 2) || (input_shape.dim_size() > 3)) {
+    return; // Input tensor should have two or three dimensions.
+  }
+  size_t n_input_dims = static_cast<size_t>(input_shape.dim_size());
+  if (input_shape.dim_size() > 2) {
+    // ignore first dim (batch axis) when there are three dims.
+    --n_input_dims;
+  }
+
+  // validating 'output_shape' input
+  std::vector<int64_t> output_shape;
+  const TensorProto* temp_output_shape = ctx.getInputData(1);
+  if (temp_output_shape) {
+    // input as initializer
+    auto target_shape = ParseData<int64_t>(temp_output_shape);
+    for (auto val : target_shape) {
+      output_shape.push_back(val);
+    }
+  } else {
+    // Only 2D - 3D output tensors are supported
+    if (ctx.getInputType(1)->tensor_type().shape().dim(0).dim_value() < 2 ||
+        ctx.getInputType(1)->tensor_type().shape().dim(0).dim_value() > 3) {
+      return;
+    }
+  }
+
+  // validating 'dilation' attribute
+  std::vector<int64_t> dilations;
+  if (getRepeatedAttribute(ctx, "dilation", dilations)) {
+    if (dilations.size() != n_input_dims) {
+      return;
+    }
+  } else {
+    dilations.assign(n_input_dims, 1);
+  }
+
+  // validating 'kernel_shape' input
+  std::vector<int64_t> kernel_shape;
+  const TensorProto* temp_kernel_shape = ctx.getInputData(2);
+  if (temp_kernel_shape) {
+    auto target_shape = ParseData<int64_t>(temp_kernel_shape);
+    for (auto val : target_shape) {
+      kernel_shape.push_back(val);
+    }
+  } else {
+    // Only 2D - 3D output tensors are supported
+    if (ctx.getInputType(2)->tensor_type().shape().dim(0).dim_value() < 2 ||
+        ctx.getInputType(2)->tensor_type().shape().dim(0).dim_value() > 3) {
+      return;
+    }
+  }
+
+  // validating 'padding' attribute
+  std::vector<int64_t> pads;
+  if (getRepeatedAttribute(ctx, "padding", pads)) {
+    if (dilations.size() != n_input_dims) {
+      return;
+    }
+  } else {
+    pads.assign(n_input_dims, 0);
+  }
+
+  // validating 'stride' attribute
+  std::vector<int64_t> strides;
+  if (getRepeatedAttribute(ctx, "stride", strides)) {
+    if (strides.size() != n_input_dims) {
+      return;
+    }
+  } else {
+    strides.assign(n_input_dims, 1);
+  }
+
+  // inferencing final shape
+  auto final_output_shape =
+      ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+
+  // try shape inference
+  auto size_of_output = static_cast<int>(output_shape.size());
+  if (size_of_output > 0) {
+    // a new dim is always added to the output
+    final_output_shape->add_dim()->set_dim_value(1);
```

Comment:
lines 2579 to 2584 are common between if and else conditions... you can extract it out

Reply:
they are similar, not identical

Reply:
In the else condition why can't you set dim 0 to 1 same as here?
It is OK if you can infer some values as long as the rank is correct.

Reply:
you are right